File: 012.jarakirefactor/djeiva/oo/heranca_polimorf/ex1/erl/Administrativo.erl
%% construtor
%%		Administrativo(String nome, float salario, int matricula, char turno,
%%						float adicionalNoturno)

%% Atributos:
%%		char turno;
%%		float adicionalNoturno;

%% Métodos:
%%	void exibeDados()

%% Atributos da superclasse:
%%		String nome;
%%		float salario;

%% Métodos das superclasses:
%%		int getMatricula()
%%		void exibeDados()

-module('Administrativo').
-export([exibeDados/1, getMatricula/1]).

'__constructor__'(ObjectID, V_nome, V_salario, V_matricula, V_turno,
					V_adicionalNoturno) ->
	Scope = {'__constructor__', 'Administrativo'},

	st:put({Scope, "nome"}, {'String', V_nome}),
	st:put({Scope, "salario"}, {float, V_salario}),
	st:put({Scope, "matricula"}, {integer, V_matricula}),
	st:put({Scope, "turno"}, {char, V_turno}),
	st:put({Scope, "adicionalNoturno"}, {float, V_adicionalNoturno}).
%% terminar...


%% da superclasse
getMatricula(ObjectID) ->
	'Assistente':getMatricula(ObjectID).

%% sobrescreve
exibeDados(ObjectID) ->
	Scope = {exibeDados, 'Administrativo'},

	Temp_Class1 = oo:super(ObjectID),
	Temp_Class1:exibeDados(),

	st:put({Scope, "turnoExtenso"}, {'String', undefined}),
	st:put({Scope, "salarioTotal"}, {'float', 0.0}),

	case (oo:get_attribute(ObjectID, "turno") == "D") of
		true ->
			st:put({Scope, "turnoExtenso"}, {'String', "diurno"});
		_ ->
			st:put({Scope, "salarioTotal"},
				oo:get_attribute(ObjectID, "salario")
				   + oo:get_attribute(ObjectID, "adicionalNoturno")),
			st:put({Scope, "turnoExtenso"}, {'String', "noturno"})
	end,

	io:format("~s~p~n", ["Turno: ", st:get(Scope, "turnoExtenso")]),

	case (oo:get_attribute(ObjectID, "turno") == "N") of
		true ->
			io:format("~s~p~n",
				["Salario + adicional noturno: ",
				 st:get(Scope, "salarioTotal")]);

		_ ->
			no_operation
	end;

%% da superclasse Assistente
exibeDados(ObjectID) ->
	'Assistente':exibeDados(ObjectID);

%% da superclasse Funcionario
exibeDados(ObjectID) ->
	'Funcionario':exibeDados(ObjectID).

File: java_src/oo/heranca/her4/Administrativo.java
// turno e adicional noturno
public class Administrativo extends Assistente {
	//char turno; // D - dia, N - noite
	int turno; // 1 - dia, 2 - noite
	float adicionalNoturno;

	//	public Administrativo(String nome, float salario, int matricula, char turno, float adicionalNoturno) {
	public Administrativo(String nome, float salario, int matricula, int turno, float adicionalNoturno) {
		//java:
		//super(nome, salario, matricula);

		this.nome = nome;
		this.salario = salario;
		this.matricula = matricula;
		//-----------------------------------------------

		this.turno = turno;
		this.adicionalNoturno = adicionalNoturno;
	}

	public void exibeDados() {
		//java:
		//super.exibeDados();

		System.out.println("======== DADOS ========");
		System.out.println("Nome: " + nome);
		System.out.println("Salario: " + salario);

		System.out.println("Matricula: " + matricula);
		//-----------------------------------------------

		String turnoExtenso;
		//java:
		//float salarioTotal = 0.0f;
		float salarioTotal = 0.0;

		if(turno == 1) {//'D') {
			turnoExtenso = "diurno";
		}else {
			salarioTotal = salario + adicionalNoturno;
			turnoExtenso = "noturno";
		}

		System.out.println("Turno: " + turnoExtenso);

		if(turno == 2)//'N')
			System.out.println("Salario + adicional noturno: " + salarioTotal);
	}
}